File: 213.house-robber-ii.py
#
# @lc app=leetcode id=213 lang=python3
#
# [213] House Robber II
#

# Time:  O(n)
# Space: O(1)

# 1d dp
# Time:  O(n)
# Space: O(n)

from typing import List


# @lc code=start
class Solution:
    def rob(self, nums: List[int]) -> int:
        def helper(start, end):
            if end - start < 2:
                return nums[start]

            # rob last or current house
            last, curr = 0, nums[start]
            for i in range(start + 1, end):
                last, curr = curr, max(last + nums[i], curr)
            return curr

            # 1d dp
            dp = [0] * end
            dp[start] = nums[start]
            dp[start + 1] = max(nums[start], nums[start + 1])
            for i in range(start + 2, end):
                dp[i] = max(nums[i] + dp[i - 2], dp[i - 1])
            return dp[-1]

        if len(nums) < 2:
            return 0 if not nums else nums[0]
        return max(helper(0, len(nums) - 1), helper(1, len(nums)))


# @lc code=end

File: 213.house-robber-ii.js
/*
 * @lc app=leetcode id=213 lang=javascript
 *
 * [213] House Robber II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {
  const helper = (start, end) => {
    if (end - start < 2) {
      return nums[start];
    }
    const dp = Array(end).fill(0);
    dp[start] = nums[start];
    dp[start + 1] = Math.max(nums[start], nums[start + 1]);
    for (let i = start + 2; i < end; ++i) {
      dp[i] = Math.max(nums[i] + dp[i - 2], dp[i - 1]);
    }
    return dp[dp.length - 1];
  };
  if (nums.length < 2) {
    return nums.length ? nums[0] : 0;
  }
  return Math.max(helper(0, nums.length - 1), helper(1, nums.length));
};
// @lc code=end
